GVRf/Framework/backend_daydream/src/main/jni/daydream_renderer.cpp

```diff
@@ -0,0 +1,252 @@
+        gvr::Mat4f result;
+        for (int i = 0; i < 4; ++i) {
+            for (int j = 0; j < 4; ++j) {
+                result.m[i][j] = 0.0f;
+                for (int k = 0; k < 4; ++k) {
+                    result.m[i][j] += matrix1.m[i][k] * matrix2.m[k][j];
+                }
+            }
+        }
+        return result;
+    }
+
+
+    static gvr::Rectf ModulateRect(const gvr::Rectf &rect, float width,
+                                   float height) {
+        gvr::Rectf result = {rect.left * width, rect.right * width,
+                             rect.bottom * height, rect.top * height};
+        return result;
+    }
+
+    static gvr::Recti CalculatePixelSpaceRect(const gvr::Sizei &texture_size,
+                                              const gvr::Rectf &texture_rect) {
+        const float width = static_cast<float>(texture_size.width);
+        const float height = static_cast<float>(texture_size.height);
+        const gvr::Rectf rect = ModulateRect(texture_rect, width, height);
+        const gvr::Recti result = {
+                static_cast<int>(rect.left), static_cast<int>(rect.right),
+                static_cast<int>(rect.bottom), static_cast<int>(rect.top)};
+        return result;
+    }
+
+
+    static void CheckGLError(const char *label) {
+        int gl_error = glGetError();
+        if (gl_error != GL_NO_ERROR) {
+            LOGW("GL error @ %s: %d", label, gl_error);
+            // Crash immediately to make OpenGL errors obvious.
+            abort();
+        }
+    }
+
+    static gvr::Sizei HalfPixelCount(const gvr::Sizei &in) {
+        // Scale each dimension by sqrt(2)/2 ~= 7/10ths.
+        gvr::Sizei out;
+        out.width = (7 * in.width) / 10;
+        out.height = (7 * in.height) / 10;
+        return out;
+    }
+
+}  // namespace
+
+DaydreamRenderer::DaydreamRenderer(JNIEnv &env, jclass clazz,
+                                   gvr_context *gvr_context)
+        : gvr_api_(gvr::GvrApi::WrapNonOwned(gvr_context)),
+          scratch_viewport_(gvr_api_->CreateBufferViewport()) {
+    jclass rendererClass = env.GetObjectClass(clazz);
+    rendererObject_ = env.NewGlobalRef(clazz);
+    onDrawEyeMethodId_ = env.GetMethodID(rendererClass, "onDrawEye", "(I)V");
+    env.DeleteLocalRef(rendererClass);
+}
+
+DaydreamRenderer::~DaydreamRenderer() {
+}
+
+void DaydreamRenderer::InitializeGl() {
+    gvr_api_->InitializeGl();
+
+    // Because we are using 2X MSAA, we can render to half as many pixels and
+    // achieve similar quality.
+    render_size_ = HalfPixelCount(gvr_api_->GetMaximumEffectiveRenderTargetSize());
+    std::vector <gvr::BufferSpec> specs;
+    specs.push_back(gvr_api_->CreateBufferSpec());
+    specs.push_back(gvr_api_->CreateBufferSpec());
+
+    specs[0].SetColorFormat(GVR_COLOR_FORMAT_RGBA_8888);
+    specs[0].SetDepthStencilFormat(GVR_DEPTH_STENCIL_FORMAT_DEPTH_16);
+    specs[0].SetSize(render_size_);
+    specs[0].SetSamples(2);
```

Comment:
We can deal with this in a subsequent pull request but shouldn't this value be coming from the settings in gvr.xml?